app/domain/authentication/authn_azure/application_identity.rb

```diff
@@ -0,0 +1,132 @@
+    # `some-namespace/service-account/some-value` is a valid application identity
+    # and will pass the validation here.
+    # If the Azure-related annotations in the host defined in Conjur has Azure-specific
+    # annotations that do not match the supplied JWT token, then it will fail the
+    # validation of ValidateApplicationIdentity.
+    class ApplicationIdentity
+
+      def initialize(role_annotations:, service_id:)
+        @role_annotations = role_annotations
+        @service_id       = service_id
+
+        validate
+      end
+
+      def constraints
+        @constraints ||= {
+            subscription_id:          constraint_value("subscription-id"),
+            resource_group:           constraint_value("resource-group"),
+            user_assigned_identity:   constraint_value("user-assigned-identity"),
+            system_assigned_identity: constraint_value("system-assigned-identity"),
+        }.compact
+      end
+
+      private
+
+      # validate that the application identity is defined correctly
+      def validate
+        # check if annotations present are part of the permitted annotations for authn-azure
+        validate_permitted_annotations
+
+        # validate that a constraint exists on the required annotations
+        validate_required_constraints_exist
+        validate_constraint_combinations
+      end
+
+      def validate_permitted_annotations
+        validate_prefixed_permitted_annotations("authn-azure/")
+        validate_prefixed_permitted_annotations("authn-azure/#{@service_id}/")
+      end
+
+      # check if annotations with prefix is part of the permitted list
+      def validate_prefixed_permitted_annotations prefix
+        Rails.logger.debug(Log::ValidatingAnnotationsWithPrefix.new(prefix))
+
+        prefixed_annotations(prefix).each do |annotation|
+          annotation_name = annotation[:name]
+          next if prefixed_permitted_constraints(prefix).include?(annotation_name)
+          raise Err::ConstraintNotSupported.new(annotation_name.gsub(prefix, ""), permitted_constraints)
+        end
+      end
+
+      def prefixed_annotations prefix
+        @role_annotations.select do |a|
+          annotation_name = a.values[:name]
+
+          annotation_name.start_with?(prefix) &&
+              # verify we take only annotations from the same level
+              annotation_name.split('/').length == prefix.split('/').length + 1
+        end
+      end
+
+      # add prefix to all permitted constraints
+      def prefixed_permitted_constraints prefix
+        permitted_constraints.map { |k| "#{prefix}#{k}" }
+      end
+
+      def validate_required_constraints_exist
+        validate_constraint_exists :subscription_id
+        validate_constraint_exists :resource_group
+      end
+
+
+      def validate_constraint_exists constraint
+        raise Err::MissingAnnotationConstraint.new(constraint) unless constraints[constraint]
```

Comment:
Does the error name need to include the word `annotation`?

Reply:
I think it makes it more clear that we are checking for constraints for Azure in the annotations. 
The msg is reads: "Host does not have {0} constraint defined in its annotations"

If we want to reuse this for other authenticators, we can make it more general. WDYT?
Maybe `Host does not have the required constraint: {0-constraint} defined in policy`?

Reply:
1. Let's not use `Host`
2. What about `Role does not have the required constraint: {0-constraint}` without stating that it's in the annotations?

Reply:
Depends if your suggestion is to make it more general so we can reuse this for other authenticators?

Reply:
What I want is to not use annotations in case the application identity will be implemented somewhere else in the future.

Reply:
updated